### PR TITLE
Fix smarthighlight

### PR DESCRIPTION
**Pull request by Cyrillev91**
Hi,
With notepad++ it is possible to select the entire word when double click on word, all words selected are highlight (smart Highlight).
But code to select word by double click is not the same as smart highlight feature.
My commit Fix smarthighlight.

more explain :
http://sourceforge.net/p/notepad-plus/patches/652/

Cyrillev

### CODE REVIEW

**Comment by Cyrillev91** [comment]
hi,
In npp you can select a word by double click.
A word contain only WordChars (A, B, ...Z, a, ..., z, and many other : use scintilla function SCI_GETWORDCHARS for complet list) not contain PunctuationChars (!"#$%&'()*+,-./:;<=>?@[]^`{|}~DEL) and not contain WhitespaceChar (SOH STX ETX EOT ENQ ACK BEL BS HT VT FF SO SI DLE DC1 DC2 DC3 DC4 NAK SYN ETB CAN EM SUB ESC FS GS RS US SPACE)
To obtain list of char use scintilla function : SCI_GETWORDCHARS , SCI_GETPUNCTUATUATIONCHARS and GET_GETWHITHSPACE)

In smarthighlight the punctuationChar and WhiteSpaceChar are hardcode don't use SCI_GETxxxxxCHARS function) and forgets grave accent.

It is also possible to change the lists with setWordChars, setPunctuationChars or setWhiteChars (with Python script plugins for example)

This Fix #187 use scintilla function in smarthighlight function.
it 's OK with grave accent : http://sourceforge.net/p/notepad-plus/bugs/5339/
And it's OK if you use setWordChars to add or remove char.

Cyrillev91

**Comment by Cyrillev91** [comment]
smarthighlight_test

**Comment by donho** [comment]
@Cyrillev91 It's interesting your modification, but I don't see any change and advantage for an end-user. Care to elaborate it?

**Comment by Eldaw** [comment]
@Cyrillev91 Now that this has been merged into v6.8.4., how can I get the smart highlighting to work like in the last column of your test? Right now, the smart highlighting is still working like it always has, and if I select END-OF-FILE then it doesn't highlight other instances of END-OF-FILE.

**Comment by Cyrillev91** [comment]
@Eldaw Hi, 
1. install python script : http://npppythonscript.sourceforge.net/
2. create script with python script (plugings >> Python script >> New scrip) with :
```python
def extendWordChar(args):
    word_chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_$-"
    for i in range(128,256):
        word_chars += chr(i)
    editor.setWordChars(word_chars)
notepad.callback(extendWordChar, [NOTIFICATION.BUFFERACTIVATED, NOTIFICATION.READY])
```
3. save script
4. run script
4. close and re-open file or change active tab file
5. now you can select by double-click "END-OF-FILE" and all "END-OF-FILE" are highlight



@donho thank you for adding my fix.
I use it with a specific language with function name or variable name allow "-" in name.
Example with MY-FUNCTION procedure or L-RES variable :

```
 011000 DEFINE, 'MY-FUNCTION', PROCEDURE                            $
 
        DECLARE, VARIABLE, 'L-RES'      IS INTEGER                  $


 011005 PERFORM, 'POWER-ON'                                         $

C-------------------------------------------------------------------$     

        PERFORM, 'TEST-XXXX'   RESULT ('L-RES')                     $

        IF, 'L-RES' EQ 0, THEN                                      $
            CALCULATE, 'FATAL' = TRUE                               $
            PERFORM, 'INIT-FAIL'                                    $
        END, IF                                                     $
 011505 PERFORM, 'POWER-OFF'                                        $
 011999 END, 'MY-FUNCTION'                                          $
 
```
Cyrillev
npp_smarthighlight

File: PowerEditor/src/ScitillaComponent/SmartHighlighter.cpp [hunk 1]
```diff
@@ -56,27 +56,34 @@ void SmartHighlighter::highlightView(ScintillaEditView * pHighlightView)
 	char * text2Find = new char[textlen];
 	pHighlightView->getSelectedText(text2Find, textlen, false);	//do not expand selection (false)
 
+	
+	//GETWORDCHARS for isQualifiedWord2() and isWordChar2()
+	int listCharSize = pHighlightView->execute(SCI_GETWORDCHARS, 0, 0);
+	char *listChar = new char[listCharSize+1];
+	pHighlightView->execute(SCI_GETWORDCHARS, 0, (LPARAM)listChar);
```
Comment (chcg) [inline PowerEditor/src/ScitillaComponent/SmartHighlighter.cpp]: String returned by SCI_GETWORDCHARS from scintilla is not null terminated, so check for strlen in isWordChar() below on listChar is dangerous

- seen with MS Application Verifier on x64

**Comment by Cyrillev91** [comment]
I agree with you

**Pull request closed without merging**